DEVELOPER_GUIDE.md

```diff
@@ -1,29 +1,44 @@
 # DEVELOPER GUIDE
 
+### Extensions
 
+This library defines many extensions for different types.
+Some of these functions are conceptually connected, in that they represent an abstraction and sometimes they have indeed a generic version of the function which doesn't require typing the type as module prefix.
+Note that in these cases, normally the names are the same but this is not a rule, there are cases where the name differs to avoid collisions or confusion with other functions, for example:
 
-### Definition of a type class
+ - `map` is a generic function, and we have many types implementing it, but for dictionaries this correspond to `mapValues`
+ - `min` is a non-generic function operating in collections, but its generic counterpart is `minimum` to avoid collision with the built-in `min` function (minimum between two values).
 
-The concept of type class is a type system construct enabling ad-hoc polymorphism.
+### Definition of abstraction
 
-in F#+, due to the constraint of targetting CIL, which doesn't have first class support for type class definition and implementation, those are not defined as F# types; instead, they are referred to in generic type parameters in the various type signatures that forms the idioms of this library implementation.
+The abstractions represented in the Control namespace of this library are a set of types that in fact represents generic functions, we call these types and we'll refer to as "Invokables".
 
-Due to that, they don't carry any effect on implementation as they are just arbitrary labels for the generic type parameters, useful to the developers and users of the library.
+These Invokables are organized in such a way that resemble typeclasses, concepts or traits in other languages where they have a type system construct enabling higher kinds and ad-hoc polymorphism.
 
-The concrete implementation of them on the other hand is done, minimally, by defining a concrete type per function.
+In F#+, due to the constraint of targeting CIL which doesn't have first class support for Higher Kinds, we can't enforce that a specific type parameter belongs to a specific abstraction. So generic type parameters are normally given the name of the abstraction, in the various type signatures that forms the idioms of this library implementation.
```

Comment:
I wonder if we could add code example for named generic parameter. Discussed concept are highly abstract, and it could help make a connection between high-level concepts and their implementation in F#+ codebase

Reply:
Definitely, we should add snippets otherwise it's too blah-blah guide, feel free to suggest concrete code fragments to include here.